hazelcast/src/hazelcast/client/sql.cpp

```diff
@@ -0,0 +1,904 @@
+          ie.get_message(),
+          boost::none,
+          exc);
+    }
+}
+
+boost::optional<member>
+query_utils::member_of_same_larger_version_group(
+  const std::vector<member>& members)
+{
+    // The members should have at most 2 different version (ignoring the patch
+    // version). Find a random member from the larger same-version group.
+
+    // we don't use 2-element array to save on litter
+    boost::optional<member::version> version0;
+    boost::optional<member::version> version1;
+    size_t count0 = 0;
+    size_t count1 = 0;
+
+    for (const auto& m : members) {
+        if (m.is_lite_member()) {
+            continue;
+        }
+        auto v = m.get_version();
+        if (!version0 || *version0 == v) {
+            version0 = v;
+            ++count0;
+        } else if (!version1 || *version1 == v) {
+            version1 = v;
+            ++count1;
+        } else {
+            throw exception::runtime(
+              "query_utils::member_of_same_larger_version_group",
+              (boost::format(
+                 "More than 2 distinct member versions found: %1% , %2%") %
+               version0 % version1)
+                .str());
+        }
+    }
+
+    assert(count1 == 0 || count0 > 0);
+
+    // no data members
+    if (count0 == 0) {
+        return boost::none;
+    }
+
+    size_t count;
+    member::version version;
+    if (count0 > count1 || (count0 == count1 && *version0 > *version1)) {
+        count = count0;
+        version = *version0;
+    } else {
+        count = count1;
+        version = *version1;
+    }
+
+    // otherwise return a random member from the larger group
+    static thread_local std::mt19937 generator;
+    std::uniform_int_distribution<int> distribution(0, count - 1);
+    auto random_member_index = distribution(generator);
+    for (const auto& m : members) {
+        if (!m.is_lite_member() && m.get_version() == version) {
+            random_member_index--;
+            if (random_member_index < 0) {
+                return m;
+            }
+        }
+    }
+
+    throw exception::runtime("query_utils::member_of_same_larger_version_group",
+                             "should never get here");
+}
+
+} // namespace impl
+
+sql_result::sql_result(
+  spi::ClientContext* client_context,
+  sql_service* service,
+  std::shared_ptr<connection::Connection> connection,
+  impl::query_id id,
+  int64_t update_count,
+  boost::optional<std::vector<sql_column_metadata>> columns_metadata,
+  boost::optional<sql_page> first_page,
+  boost::optional<bool> is_infinite_rows,
+  int32_t cursor_buffer_size)
+  : client_context_(client_context)
+  , service_(service)
+  , connection_(std::move(connection))
+  , query_id_(id)
+  , update_count_(update_count)
+  , first_page_(std::move(first_page))
+  , is_infinite_rows_(is_infinite_rows)
+  , iterator_requested_(false)
+  , closed_(false)
+  , cursor_buffer_size_(cursor_buffer_size)
+{
+    if (columns_metadata) {
```

Comment:
fixed the first comment. see the latest changes. 

2nd comment: We do not need the logic in Java because, the user will get exception when the user does future.get() and user can not have the sql_result object in case of the error. There is no possibility of not having row metadata and the update count will be set in case of no errors. does this make sense?

Reply:
Not talking about the error case.


A rowset or an update count result can be returned from the server. A rowset result has rowMetadata set. Update count result has rowMetadata is `null` and updateCount should be -1. Actually server side sends -1 in that case, it is safer to set it to -1 manually just like Java client [does](https://github.com/hazelcast/hazelcast/blob/99b029b2582ce4a88184d49140f4c09d0f4f677e/hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java#L114) because server side can change.

See [SqlExecuteCodec.encodeResponse](https://github.com/hazelcast/hazelcast/blob/67a8b58d0e445230ed9ff9805f109bd2bd9063bc/hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlExecuteMessageTask.java#L73)

Reply:
Ok, fixed 44138ba6d61c91b53cf300c8e3adcd15949f9e50